src/lib/utils/error-logger.js

```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+const logger = {};
+
+const getLogData = (err) => {
+    let logData;
+    if (err.stack) {
+        logData = err.stack;
+    } else if (err.message) {
+        logData = err.message;
+    } else if (err.toString instanceof Function) {
+        logData = err.toString();
+    } else {
+        logData = err;
+    }
+
+    return logData;
+};
+
+module.exports.register = (server, options, next) => {
+
+    //logs error
+    logger.logError = (tags,err) => {
+        server.log(['error'].concat(tags), getLogData(err));
+    };
+
+    //logs and throws error
+    logger.logAndThrowError = (tags,err) => {
```

Comment:
Hey @dylancwood!  Thanks for the response.  I've taken some time to digest your rebuttle, and I'd like to keep the convo going, as I still think we should collectively reconsider.

> you advocate that utility B be written such that it relies on its consumer (utility A) to perform logging 

This is not the intent of my example.  In my example, `B` may not be even be coins authored (or `B` uses some coins-external package internally), and therefore would have no means to tap into something like `logAndThrow`. If `A` depends on `B`, and `A.someMethod` is at the root of the stack, `A.someMethod` should only log (of the two), assuming it's desired to be fault tolerant.  I would advocate that arbitrary logging creates bloat and paints an unclear wholistic picture of what's happening in the logs.  Only where the error is handled (if it is handled at all) should we log.  Suppose the stack looks like:
```js
server.listen
a.go
b.go
c.go
```
and c errors, then you log and throw in a, b, and c.  One issue now looks like **three**, biasing error stats and possible assign-ability to the actual issue at hand.  Only "one-thing" should report a summary of the error, otherwise it just gets messy.

> utility B should log information that is useful in debugging that error

I still kindly disagree.  If B is logging, and then in A's error handler it also logs, you've now done duplicate error logging/handling.  I think that Error handling and logging should be owned by lowest level possible in the stack.  If B throws, it still has all of the meta-data desired, including stack, wrapped up in the error.  If A wants to make note, fine, it can log, but not if it's just going to rethrow again--the root/global error handler should _already be writing unrecoverable/unhandleable error content_.  Under this scheme, we have a single record of a mishap, with all of the pertinent meta-data.  no need to redundantly log.  To say that such a utility is needed I think indirectly infers that the provisions of the built-ins are insufficient.

In principle, I understand the concerns, but I think we should make an effort to define when we should and when we should not log.  I think we would all agree "log-at-will" whenever we detect an error state may not be a great pattern, as it fills the logs with much non-actionable content.  I think a policy of "handle errors where the system/route _actually can adapt to errors_" is a mandatory step, which we already make efforts to follow.  However, code that does something like

```js
someAsyncThing()
.catch(function(err) {
   log(err);
   throw err;
});
```
provides no value, as the `err` will otherwise have already been passed up to the root handler, and have all of the same metadata, which should be logged by default.

In your provided example, providing a `401` is _not a server application error_--it's a client error.  The server is not in an erroring state.  It seems at face value to be dangerous to declare the server application in an error state when it is indeed working as expected.

Phew.  Sorry for the **huge** response.  I'm happy to continue on github, slack, or meeting.